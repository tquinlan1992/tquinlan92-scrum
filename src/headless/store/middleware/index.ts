// import saveToLocalStorageState from './saveToLocalStorageState';
import thunk from 'redux-thunk';
import { routerÂ } from './router';
import logger from 'redux-logger';

export const middleware = [
    // saveToLocalStorageState,
    router,
    thunk,
    logger
];
