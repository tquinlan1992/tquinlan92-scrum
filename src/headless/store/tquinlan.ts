import { Dictionary, mapValues, omit, merge } from "lodash";
import { Reducer, AnyAction } from "redux";
import { makeSimpleReducer, ActionCreator } from "tquinlan92-typescript-redux-utils";

export function getActions<T extends { [key: string]: {actions?: Dictionary<any>} }>(creators: T): { [P in keyof T]: T[P]['actions'] } {
    return mapValues(creators, "actions") as { [P in keyof T]: T[P]['actions'] };
} 

export function makeNestedSimpleReducerSimpleActions<AppState>(state: any) {
    const actionsReducers = mapValues(state, (value, key) => {
        return makeSimpleReducer(key, omit(value, 'actions'));
    });//{ [P in keyof T]: T[P]['actionCreator'] 
    const reducers = mapValues(actionsReducers, 'reducer');
    const actions = getActions(actionsReducers);

    return {
        actions,
        reducers
    } as {
        reducers: {
            [P in keyof AppState]: Reducer<AppState[P], AnyAction>
        };
        actions: {
            [P in keyof AppState]: {
                [A in keyof AppState[P]]: ActionCreator<AppState[P][A]>
            } & {
                reset: ActionCreator<null>;
                setAll: ActionCreator<AppState[P]>;
                set: ActionCreator<Partial<AppState[P]>>;
            } & {
                [A in keyof AppState[P]]: AppState[P][A]
            }
        };
    };
}

export function makeNestedSimpleStore<State, ThunkActions>(state: State, thunkActions?: ThunkActions) {
    const { actions: actions1, reducers: reducers2 } = makeNestedSimpleReducerSimpleActions<State>(state);
    const actionsWithThunks = merge(actions1, thunkActions);
    return {
        actions: actionsWithThunks,
        reducers: reducers2
    }
}