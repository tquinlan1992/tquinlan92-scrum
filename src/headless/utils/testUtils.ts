export function mockPouchDB() {
    jest.mock('../database/core/pouch', () => {
        return {};
    });
}
