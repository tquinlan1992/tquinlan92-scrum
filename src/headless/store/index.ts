import { createStore, applyMiddleware, AnyAction, Store, combineReducers } from "redux";
import { middleware } from './middleware';
import { composeWithDevTools } from 'redux-devtools-extension';
import { routerReducer, RouterState } from "react-router-redux";
import { TicketListState } from '@components/TicketList/redux';
import { AddTicketState, addTicketDialogReducer } from '@components/AddTicketDialog/redux';
import { LoadingState, loadingReducer } from '@components/Loading/redux';
import { AppState } from '../store';
import { ticketListReducer } from '../../components/TicketList/redux';
import { ThunkAction } from "redux-thunk";
import { CloseSprintDialogState, closeSprintDialogReducer } from '@components/CloseSprintDialog/redux';


const appReducer = combineReducers<AppState>({
    routing: routerReducer,
    addTicket: addTicketDialogReducer,
    ticketList: ticketListReducer,
    loading: loadingReducer,
    closeSprintDialog: closeSprintDialogReducer
});

declare module 'redux' {
    export type GenericStoreEnhancer = any;
}

export type AppThunkAction = ThunkAction<void, AppState, void, AnyAction>;

export interface Doc {
    edit: string;
}

export interface VoiceToCode {
    text: string;
}

export interface AddTicketState {
    storyPoint: null | number;
    description: string;
    title: string;
}

export interface AppState {
    routing: RouterState;
    ticketList: TicketListState;
    addTicket: AddTicketState;
    loading: LoadingState;
    closeSprintDialog: CloseSprintDialogState;
}

const reduxStore: Store<AppState, AnyAction> = createStore<AppState, AnyAction, void, void>(appReducer, {}, composeWithDevTools<any, {}>(applyMiddleware(...middleware)));

export { reduxStore };
