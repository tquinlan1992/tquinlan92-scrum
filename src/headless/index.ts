import { reduxStore } from './store';

export { reduxStore };
