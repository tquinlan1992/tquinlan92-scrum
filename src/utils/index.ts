import * as _ from 'lodash';
import { connect, MapStateToPropsParam, MergeProps } from 'react-redux';
import { AppState } from '@headless/store';
import { ThunkDispatch } from 'redux-thunk';
import { AnyAction } from 'redux';
import { Theme } from '@material-ui/core';
import { CSSProperties } from '@material-ui/core/styles/withStyles';
import { history } from '@src/headless/store/middleware/router';

type PropsMap<M, K extends keyof M> = {[P in K]: M[P]};

export function pick<M extends object, K extends keyof M>(obj: M, ...props: K[]): PropsMap<M, K> {
    // @ts-ignore: TS2352: Type 'PartialDeep<M>' cannot be converted to type 'PropsMap<M, K>'.
    return _.pick<M>(obj, props) as PropsMap<M, K>;
}

export type Diff<T extends string, U extends string> = ({ [P in T]: P } & { [P in U]: never } & { [x: string]: never })[T];

export type AppStateThunkDispatch = ThunkDispatch<AppState, undefined, AnyAction>;

type CreateStylesReturnTheme<C extends string> = Record<C, CSSProperties>;

type GetThemeCallback<C extends string> = (theme: Theme) => CreateStylesReturnTheme<C>;

export function getTheme<C extends string>(getThemeCallback: GetThemeCallback<C>) {
    return (theme: Theme) => getThemeCallback(theme);
}

export function navigateTo(path: string) {
    history.push(path)
}
