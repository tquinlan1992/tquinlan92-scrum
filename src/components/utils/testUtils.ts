import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { AnyAction, isType, ActionCreator } from 'typescript-fsa';
import { AppState } from '../../headless/store';

export function getAnyJestFn() {
    return jest.fn() as any;
}

interface AnyActionPayload extends AnyAction {
    payload: any;
}

export function expectActionWithPayload(actionToTest: AnyActionPayload, expectedActionCreator: ActionCreator<any>, expectedPayload?: any) {
    expect(isType(actionToTest, expectedActionCreator)).toBe(true);
    expect(actionToTest.payload).toEqual(expectedPayload);
}


const middlewares = [thunk];
export const mockStore = configureStore(middlewares);

export function getMockStore(state: Partial<AppState>) {
    return mockStore(state);
}

export function expectCalledOnceWith(mockFunction: jest.Mock<{}>, calledWith?: any) {
    expect(mockFunction).toHaveBeenCalledTimes(1);
    const expectedParams = calledWith ? [calledWith] : [];
    expect(mockFunction.mock.calls[0]).toEqual(expectedParams);
}

export function mockClearAll(mockFunctions: jest.Mock<{}>[]) {
    mockFunctions.forEach(mockFunction => {
        mockFunction.mockClear();
    })
}