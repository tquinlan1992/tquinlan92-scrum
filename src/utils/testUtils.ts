import { Ticket, Docs } from '../headless/database/PouchWrapper';
import pouchDBMemoryAdapter from 'pouchdb-adapter-memory';
import PouchDB from 'pouchdb-browser';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { AnyAction, isType, ActionCreator } from 'typescript-fsa';
import { AppState } from '../headless/store';
import { DeepPartial } from 'ts-essentials';

const pouchdbFind = require('pouchdb-find');
PouchDB.plugin(pouchDBMemoryAdapter);
PouchDB.plugin(pouchdbFind);

export async function getMemoryPouchDb(docs?: Array<Docs['Ticket']>) {
    const db = new PouchDB<Docs['Ticket']>('mydb', { adapter: 'memory' });
    if (docs) {
        docs.forEach(async doc => {
            try {
            await db.put(doc);
            } catch(e) {
                console.log('e', e);
            }
        })
    }
    return db;
}



export function mockPouchDB() {
    jest.mock('@headless/database/pouch', () => {
        return {};
    });
}

export function mockPouch(dbMethods: object) {
    jest.mock('@headless/database/pouch', () => {
        return {

        }
    })
    const mockPouchRequire = require('@headless/database/pouch');
    const mockGetRemoteDb = jest.fn(() => {
        return new Promise(resolve => {
            resolve(dbMethods);
        })
    });
    Object.assign(
        mockPouchRequire, 
        {
            getRemoteDb: mockGetRemoteDb 
        }, 
        dbMethods
    )
    return mockPouchRequire;
}

export function getAnyJestFn() {
    return jest.fn() as any;
}

interface AnyActionPayload extends AnyAction {
    payload: any;
}

export function expectActionWithPayload(actionToTest: AnyActionPayload, expectedActionCreator: ActionCreator<any>, expectedPayload?: any) {
    expect(isType(actionToTest, expectedActionCreator)).toBe(true);
    expect(actionToTest.payload).toEqual(expectedPayload);
}


const middlewares = [thunk];
export const mockStore = configureStore<Partial<AppState>, any>(middlewares);

export function getMockState(state: DeepPartial<AppState>) {
    return state as AppState;
}

export function getMockStore(state: DeepPartial<AppState>) {
    return mockStore(state as AppState);
}

export function expectCalledOnceWith(mockFunction: jest.Mock<{}>, calledWith?: any) {
    expect(mockFunction).toHaveBeenCalledTimes(1);
    const expectedParams = calledWith ? [calledWith] : [];
    expect(mockFunction.mock.calls[0]).toMatchObject(expectedParams);
}

export function mockClearAll(mockFunctions: jest.Mock<{}>[]) {
    mockFunctions.forEach(mockFunction => {
        mockFunction.mockClear();
    })
}

export function getPartialState(appState: DeepPartial<AppState>) {
    return appState as AppState;
}