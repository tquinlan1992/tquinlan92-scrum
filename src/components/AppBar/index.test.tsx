import { AppBar, mapStateToProps } from './';
import { shallow } from 'enzyme';
import * as React from 'react';
import { AppState } from '../../headless/store';
import { DeepPartial } from 'ts-essentials';
import { getMockState } from '@src/utils/testUtils';

describe('test snapshot of Loading component', () => {
    it('should render the loading screen when loading is true', () => {
        const props = {
            classes: {
                appBar: 'appBar',
                center: 'center'
            },
            loading: true,
            openAddTicketDialog: jest.fn(),
            history: {} as any,
            location: {
                path: '/feed'
            } as any,
            match: {} as any
        };
        const result = shallow(<AppBar {...props} />);
        expect(result).toMatchSnapshot();
    });
    it('should render the app when loading is false', () => {
        const props = {
            classes: {
                appBar: 'appBar',
                center: 'center'
            },
            loading: false,
            openAddTicketDialog: jest.fn(),
            history: {} as any,
            location: {
                path: '/feed'
            } as any,
            match: {} as any
        };
        const result = shallow(<AppBar {...props} />);
        expect(result).toMatchSnapshot();
    });
});

describe('mapStateToProps', () => {
    const state = getMockState({
        loading: {
            value: true
        }
    });
    const props = mapStateToProps(state);
    it('should have valid props', () => {
        expect(props).toEqual({
            loading: true
        })
    })
})
