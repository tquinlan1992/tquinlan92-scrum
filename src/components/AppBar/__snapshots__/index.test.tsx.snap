// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test snapshot of Loading component should render the app when loading is false 1`] = `
<WithStyles(AppBar)
  className="appBar"
  color="secondary"
  position="fixed"
>
  <WithStyles(Grid)
    container={true}
    style={
      Object {
        "height": "67px",
      }
    }
  >
    <WithStyles(Grid)
      className="center"
      item={true}
      style={
        Object {
          "cursor": "pointer",
          "margin": "auto",
          "width": "10%",
        }
      }
    >
      <img
        src="logo.png"
        style={
          Object {
            "height": "55px",
            "margin": "auto",
          }
        }
      />
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
      style={
        Object {
          "width": "80%",
        }
      }
    >
      <WithStyles(Tabs)
        centered={true}
        indicatorColor="primary"
        onChange={[Function]}
        textColor="primary"
        value={null}
      >
        <WithStyles(Tab)
          disabled={true}
          label="Closed"
        />
        <WithStyles(Tab)
          disabled={true}
          label="Open"
        />
      </WithStyles(Tabs)>
    </WithStyles(Grid)>
    <WithStyles(Grid)
      className="center"
      item={true}
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#00D036",
          "cursor": "pointer",
          "height": "67px",
          "width": "10%",
        }
      }
    >
      <WithStyles(Typography)
        style={
          Object {
            "color": "white",
            "fontSize": "17px",
            "fontWeight": "bold",
            "lineHeight": "47px",
          }
        }
      >
        + New Task
      </WithStyles(Typography)>
    </WithStyles(Grid)>
  </WithStyles(Grid)>
</WithStyles(AppBar)>
`;

exports[`test snapshot of Loading component should render the loading screen when loading is true 1`] = `
<WithStyles(AppBar)
  className="appBar"
  color="secondary"
  position="fixed"
>
  <WithStyles(Grid)
    container={true}
    style={
      Object {
        "height": "67px",
      }
    }
  >
    <WithStyles(Grid)
      className="center"
      item={true}
      style={
        Object {
          "cursor": "pointer",
          "margin": "auto",
          "width": "10%",
        }
      }
    >
      <img
        src="logo.png"
        style={
          Object {
            "height": "55px",
            "margin": "auto",
          }
        }
      />
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
      style={
        Object {
          "width": "80%",
        }
      }
    >
      <WithStyles(Tabs)
        centered={true}
        indicatorColor="primary"
        onChange={[Function]}
        textColor="primary"
        value={null}
      >
        <WithStyles(Tab)
          disabled={true}
          label="Closed"
        />
        <WithStyles(Tab)
          disabled={true}
          label="Open"
        />
      </WithStyles(Tabs)>
    </WithStyles(Grid)>
    <WithStyles(Grid)
      className="center"
      item={true}
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#00D036",
          "cursor": "pointer",
          "height": "67px",
          "width": "10%",
        }
      }
    >
      <WithStyles(Typography)
        style={
          Object {
            "color": "white",
            "fontSize": "17px",
            "fontWeight": "bold",
            "lineHeight": "47px",
          }
        }
      >
        + New Task
      </WithStyles(Typography)>
    </WithStyles(Grid)>
  </WithStyles(Grid)>
</WithStyles(AppBar)>
`;
